File: rebound_rust/porttest/archive_test.c
/* archive_test.c — C side of the Simulationarchive cross-language
 * round-trip verification.
 * Usage: archive_test <integrator> write     — run 3x100 steps, saving
 *          a snapshot after each 100 to archive_c_<integrator>.bin;
 *          dump the final (300 step) state bits to archive_state_c.txt.
 *        archive_test <integrator> continue  — load snapshot 1 (the
 *          200-step state) from archive_rust_<integrator>.bin, run 100
 *          more steps, dump state bits to archive_state_c.txt.
 * Part of the rebound_rs port verification. GPL-3.0-or-later. */
#include "rebound.h"
#include "integrator_whfast.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static unsigned long long bits(double x){
    unsigned long long u; memcpy(&u,&x,8); return u;
}

static void setup_particles(struct reb_simulation* r){
    struct reb_particle star = {0};
    star.m = 1.0;
    reb_simulation_add(r, star);
    struct reb_particle planet = {0};
    planet.m = 1e-3; planet.x = 1.6; planet.vy = 0.5;
    reb_simulation_add(r, planet);
    struct reb_particle moon = {0};
    moon.m = 1e-7; moon.x = 1.7; moon.vy = 0.6; moon.z = 0.01; moon.vz = 0.001;
    reb_simulation_add(r, moon);
}

static void configure(struct reb_simulation* r, const char* integrator, void* state){
    if (strcmp(integrator,"whfast-usafe")==0){
        struct reb_integrator_whfast_state* wh = state;
        wh->safe_mode = 0;
    }
    (void)r;
}

static void dump_state(struct reb_simulation* r, const char* integrator){
    FILE* f = fopen("archive_state_c.txt","wb");
    fprintf(f, "integrator %s\n", integrator);
    fprintf(f, "t %016llx\n", bits(r->t));
    fprintf(f, "dt %016llx\n", bits(r->dt));
    for (size_t i=0;i<r->N;i++){
        struct reb_particle p = r->particles[i];
        fprintf(f, "%llu %016llx %016llx %016llx %016llx %016llx %016llx\n",
                (unsigned long long)i,
                bits(p.x), bits(p.y), bits(p.z),
                bits(p.vx), bits(p.vy), bits(p.vz));
    }
    fclose(f);
}

int main(int argc, char* argv[]){
    const char* integrator = argc>1 ? argv[1] : "whfast-usafe";
    const char* mode = argc>2 ? argv[2] : "write";

    const char* real_integrator = integrator;
    if (strncmp(integrator, "whfast", 6)==0) real_integrator = "whfast";

    char fname[256];

    if (strcmp(mode,"load")==0){
        /* Load snapshot 0 from the file given in argv[3], dump state. */
        const char* lfname = argc>3 ? argv[3] : "served.bin";
        struct reb_simulation* r = reb_simulation_create_from_file((char*)lfname, 0);
        if (!r){
            printf("Failed to load %s\n", lfname);
            return 1;
        }
        dump_state(r, integrator);
        printf("load done: t=%e\n", r->t);
        return 0;
    }
    if (strcmp(mode,"write")==0){
        struct reb_simulation* r = reb_simulation_create();
        void* state = reb_simulation_set_integrator(r, real_integrator);
        configure(r, integrator, state);
        r->G = 1.0;
        r->dt = 0.01;
        setup_particles(r);
        sprintf(fname, "archive_c_%s.bin", integrator);
        remove(fname);
        for (int s=0;s<3;s++){
            reb_simulation_steps(r, 100);
            reb_simulation_save_to_file(r, fname);
        }
        dump_state(r, integrator);
        printf("write done: t=%e\n", r->t);
    }else{
        sprintf(fname, "archive_rust_%s.bin", integrator);
        struct reb_simulation* r = reb_simulation_create_from_file(fname, 1);
        if (!r){
            printf("Failed to load %s\n", fname);
            return 1;
        }
        reb_simulation_steps(r, 100);
        dump_state(r, integrator);
        printf("continue done: t=%e\n", r->t);
    }
    return 0;
}

File: rebound_rust/examples/archive_test.rs
//! archive_test.rs — Rust side of the Simulationarchive cross-language
//! round-trip verification (twin of porttest/archive_test.c, with the
//! file roles swapped: writes archive_rust_<integrator>.bin, continues
//! from archive_c_<integrator>.bin, dumps to archive_state_rust.txt).
//! Part of rebound_rs, GPL-3.0-or-later.
#![allow(non_snake_case)]
// Clippy waivers. A test/example is its own crate and does not inherit
// the crate root's waivers, so they are repeated here. Same justification:
// this code mirrors the C source's idioms, and applying clippy's
// suggestions would obscure the correspondence that makes the port
// reviewable. See rebound_rust.md section 17.
#![allow(clippy::manual_clamp)] // mirrors the C's explicit min/max tests
#![allow(clippy::neg_cmp_op_on_partial_ord)]
#![allow(clippy::excessive_precision)]
#![allow(clippy::identity_op)]
#![allow(clippy::erasing_op)]
#![allow(clippy::needless_range_loop)]
#![allow(clippy::assign_op_pattern)]
#![allow(clippy::field_reassign_with_default)]
#![allow(clippy::too_many_arguments)]
#![allow(clippy::manual_range_contains)]
#![allow(clippy::manual_memcpy)]
#![allow(clippy::manual_swap)]
#![allow(clippy::manual_div_ceil)]
#![allow(clippy::manual_is_multiple_of)]
#![allow(clippy::misrefactored_assign_op)]
#![allow(clippy::neg_multiply)]
#![allow(clippy::collapsible_if)]
#![allow(clippy::collapsible_else_if)]
#![allow(clippy::needless_late_init)]
#![allow(clippy::while_let_loop)]
#![allow(clippy::unnecessary_cast)]
#![allow(clippy::ptr_arg)]
#![allow(clippy::seek_from_current)]
#![allow(clippy::drop_non_drop)]
#![allow(clippy::redundant_closure)]
#![allow(clippy::approx_constant)]
#![allow(clippy::useless_vec)]
#![allow(clippy::type_complexity)]

use rebound_rs::*;
use std::io::Write;

fn bits(x: f64) -> u64 {
    x.to_bits()
}

fn setup_particles(r: &mut reb_simulation) {
    let mut star = reb_particle::default();
    star.m = 1.0;
    reb_simulation_add(r, star);
    let mut planet = reb_particle::default();
    planet.m = 1e-3;
    planet.x = 1.6;
    planet.vy = 0.5;
    reb_simulation_add(r, planet);
    let mut moon = reb_particle::default();
    moon.m = 1e-7;
    moon.x = 1.7;
    moon.vy = 0.6;
    moon.z = 0.01;
    moon.vz = 0.001;
    reb_simulation_add(r, moon);
}

fn configure(r: &mut reb_simulation, integrator: &str) {
    if integrator == "whfast-usafe" {
        if let reb_integrator_state::whfast(ref mut wh) = r.integrator {
            wh.safe_mode = 0;
        }
    }
}

fn dump_state(r: &reb_simulation, integrator: &str) {
    let mut f = std::fs::File::create("archive_state_rust.txt").unwrap();
    writeln!(f, "integrator {}", integrator).unwrap();
    writeln!(f, "t {:016x}", bits(r.t)).unwrap();
    writeln!(f, "dt {:016x}", bits(r.dt)).unwrap();
    for i in 0..r.N {
        let p = r.particles[i];
        writeln!(
            f,
            "{} {:016x} {:016x} {:016x} {:016x} {:016x} {:016x}",
            i,
            bits(p.x),
            bits(p.y),
            bits(p.z),
            bits(p.vx),
            bits(p.vy),
            bits(p.vz)
        )
        .unwrap();
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let integrator = if args.len() > 1 { args[1].clone() } else { "whfast-usafe".to_string() };
    let mode = if args.len() > 2 { args[2].clone() } else { "write".to_string() };

    let real_integrator = if integrator.starts_with("whfast") { "whfast" } else { integrator.as_str() };

    if mode == "write" {
        let mut sim = reb_simulation_create();
        let r = &mut sim;
        reb_simulation_set_integrator(r, real_integrator);
        configure(r, &integrator);
        r.G = 1.0;
        r.dt = 0.01;
        setup_particles(r);
        let fname = format!("archive_rust_{}.bin", integrator);
        let _ = std::fs::remove_file(&fname);
        for _ in 0..3 {
            reb_simulation_steps(r, 100);
            reb_simulation_save_to_file(r, Some(&fname));
        }
        dump_state(r, &integrator);
        println!("write done: t={:e}", r.t);
    } else {
        let fname = format!("archive_c_{}.bin", integrator);
        let mut sim = match reb_simulation_create_from_file(&fname, 1) {
            Some(s) => s,
            None => {
                println!("Failed to load {}", fname);
                std::process::exit(1);
            }
        };
        let r = &mut sim;
        reb_simulation_steps(r, 100);
        dump_state(r, &integrator);
        println!("continue done: t={:e}", r.t);
    }
}
